File: 23. Build from inorder and preorder.cpp
#include <bits/stdc++.h>
using namespace std;
struct Node
{
    int data;
    Node* left;
    Node* right;
    Node(int d) : data(d), left(nullptr), right(nullptr) {}
};
Node* buildIP(vector<int>& inorder, vector<int>& preorder, int inStart, int inEnd, int& pIndex,
              unordered_map<int, int>& idx)
{
    if (inStart > inEnd)
        return nullptr;
    int rootVal = preorder[pIndex++];
    Node* root = new Node(rootVal);
    int inIndex = idx[rootVal];
    root->left = buildIP(inorder, preorder, inStart, inIndex - 1, pIndex, idx);
    root->right = buildIP(inorder, preorder, inIndex + 1, inEnd, pIndex, idx);
    return root;
}
Node* buildFromInPre(vector<int> inorder, vector<int> preorder)
{
    unordered_map<int, int> idx;
    for (int i = 0; i < inorder.size(); i++) idx[inorder[i]] = i;
    int pIndex = 0;
    return buildIP(inorder, preorder, 0, inorder.size() - 1, pIndex, idx);
}
void inorder(Node* root)
{
    if (!root)
        return;
    inorder(root->left);
    cout << root->data << " ";
    inorder(root->right);
}
int main()
{
    vector<int> inorder = {4, 2, 5, 1, 3};
    vector<int> preorder = {1, 2, 4, 5, 3};
    Node* root = buildFromInPre(inorder, preorder);
    inorder(root);
    cout << endl;
    return 0;
}

File: 23. Build from inorder and preorder.py
class Node:
    def __init__(self,data): self.data=data; self.left=None; self.right=None

def build_in_pre(inorder, preorder):
    idx={v:i for i,v in enumerate(inorder)}
    pIndex=0
    def helper(l,r):
        nonlocal pIndex
        if l>r: return None
        rootVal=preorder[pIndex]; pIndex+=1
        root=Node(rootVal)
        inIdx=idx[rootVal]
        root.left=helper(l,inIdx-1)
        root.right=helper(inIdx+1,r)
        return root
    return helper(0,len(inorder)-1)

def inorder(root):
    if not root: return
    inorder(root.left); print(root.data, end=' '); inorder(root.right)

if __name__=='__main__':
    inorder_list=[4,2,5,1,3]
    preorder=[1,2,4,5,3]
    root=build_in_pre(inorder_list, preorder)
    inorder(root); print()
